In [14]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import matplotlib.pyplot as plt

In [15]:
# Veri setini yükleyelim
file_path = 'updated_cybersecurity_attacks.csv'
data = pd.read_csv(file_path)

In [16]:
# 'Protocol' sütununu kategorik verilerden sayısal değerlere dönüştürelim
data['Protocol'] = data['Protocol'].astype('category').cat.codes
# Gerekli sütunları seçelim
features = data[['Source Port', 'Destination Port', 'Protocol', 'Packet Length']]
target = data['Traffic Type']


In [17]:

# Özellikleri ölçeklendirelim
scaler = StandardScaler()
scaled_features = scaler.fit_transform(features)

In [18]:
# Veriyi eğitim ve test setlerine bölelim
X_train, X_test, y_train, y_test = train_test_split(scaled_features, target, test_size=0.3, random_state=42)

In [19]:
# Veriyi eğitim ve test setlerine bölelim
X_train, X_test, y_train, y_test = train_test_split(scaled_features, target, test_size=0.3, random_state=42)


In [20]:
# k-NN modelini tanımlayalım
knn = KNeighborsClassifier(n_neighbors=5)

# Modeli eğitelim
knn.fit(X_train, y_train)

# Test seti üzerinde tahmin yapalım
y_pred = knn.predict(X_test)

# Model doğruluğunu hesaplayalım
accuracy = accuracy_score(y_test, y_pred)

In [21]:
print(f'k-NN Model Accuracy: {accuracy * 100:.2f}%')

# Detaylı sınıflandırma raporu
print(classification_report(y_test, y_pred))

# Karışıklık matrisi
conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

k-NN Model Accuracy: 33.13%
              precision    recall  f1-score   support

         DNS       0.33      0.46      0.39      4016
         FTP       0.33      0.33      0.33      3953
        HTTP       0.34      0.20      0.25      4031

    accuracy                           0.33     12000
   macro avg       0.33      0.33      0.32     12000
weighted avg       0.33      0.33      0.32     12000

Confusion Matrix:
[[1860 1344  812]
 [1865 1307  781]
 [1866 1356  809]]


### Değerlendirme

Sonuçlar doğruluğun %33.13 olduğunu gösteriyor, bu da sınıflandırmanın üç sınıf arasında rastgele yapıldığında beklenen doğruluğa yakın olduğunu gösterir. Precision, recall ve f1-score değerleri de sınıflandırmanın yeterince iyi olmadığını göstermekte. Bu, modelimizin geliştirilebileceği anlamına gelir.

 